### DISCUSSION TITLE
vela def init function help&improvement

### DISCUSSION THREAD

**Discussion by LinzeSu** [kubevela/kubevela#5838] [Q&A]
Hi,
I noticed there's [description ](https://github.com/kubevela/kubevela/blob/c007ad8df04ed445d5aad050cd966cdeb15173b3/references/cli/def.go#L166) of the NewDefinitionInitCommand that goes:
`However, we do not support variables in YAML file currently, which prevents users from directly feeding files like helm chart directly.`
I don't really understand. What exactly does it mean?

Also, when using `vela def init -t component --template-yaml my-object.yaml -o my-component.cue` functions to generate the cue file, the parameters are empty, which can result in many cubersome parameters definition work. Would it be nice that we add an option that allows all the parameters to be autofilled in the generated cue file?

**Comment by wonderflow** [comment]
1. The first question:


> However, we do not support variables in YAML file currently, which prevents users from directly feeding files like helm chart directly.

It means format like this which inlcude go template in it can not be supported, we can only unmarshal valid YAML:

```
{{- if .Values.serviceAccount.create -}}
apiVersion: v1
kind: ServiceAccount
metadata:
  name: {{ include "kubevela.serviceAccountName" . }}
  labels:
    {{- include "kubevela.labels" . | nindent 4 }}
  {{- with .Values.serviceAccount.annotations }}
  annotations:
  {{- toYaml . | nindent 4 }}
  {{- end }}
{{- end }}
```

2. The second question

We can not identity which fields should be parameters in this case, do you have any suggestions?

**Reply by LinzeSu** [reply]
Thanks for the explanation.
And about parameters, it depends on use case. Just some quick thought, I would suggest adding all the fileds as parameter, with the default value identical with the value defined in yaml file, so it enables both custumization and flexibility.

**Reply by wonderflow** [reply]
Which are "all" fields in the following case? It's `spec.template.spec.containers`? Or `ports` in the first container? Or the `containerPort`?

```
        apiVersion: "apps/v1"
        kind:       "StatefulSet"
        metadata: name: "web"
        spec: {
            selector: matchLabels: app: "nginx"
            replicas:    3
            serviceName: "nginx"
            template: {
                metadata: labels: app: "nginx"
                spec: {
                    containers: [{
                        name: "nginx"
                        ports: [{
                            name:          "web"
                            containerPort: 80
                        }]
                        image: "k8s.gcr.io/nginx-slim:0.8"
                        volumeMounts: [{
                            name:      "www"
                            mountPath: "/usr/share/nginx/html"
                        }]
                    }]
                    terminationGracePeriodSeconds: 10
                }
            }
            volumeClaimTemplates: [{
                metadata: name: "www"
                spec: {
                    accessModes: ["ReadWriteOnce"]
                    resources: requests: storage: "1Gi"
                    storageClassName: "my-storage-class"
                }
            }]
        }
```

**Reply by LinzeSu** [reply]
All fields and their subfieds except `apiVersion` and `kind`, including the metadata. In this case, "all" would include all the information except the first two lines.

**Reply by wonderflow** [reply]
I won't like that experience as I can take the same time removing the parameters I don't need.

**Reply by LinzeSu** [reply]
Just to make sure I made it clear. Let's say we have a YAML that defines a kind of resource that contains a lot of setups, transforming it to the cue file would require a whole bunch of setups in the parameters too. The setups here are necessary to deploy the resource so there would be no "unnecessary parameters". The parameters should contain all the setups and have the default value, so that if one use it later in as a component, they can just use the boilerplate offered or choose to override the default values.
If that is how the things could work, that would help me saving a lot of time cause some of my resource YAML files contain hundreds of lines. Also, I believe there should be a more general way of doing this, but for now this is the idea I have in mind.

**Reply by wonderflow** [reply]
You can give me an expected example using the yaml provided above for better understanding.

**Reply by LinzeSu** [reply]
Sure, it will look something like this. This demo is generated using some chat bot combining previous conversation. Some mistakes exist but I hope it's enough to express the thought.
```
template: {
    output: {
        apiVersion: parameter.apiVersion
        kind:       parameter.kind
        metadata: name: parameter.metadata.name
        spec: {
            selector: matchLabels: app: "nginx"
            replicas:    parameter.spec.replicas
            serviceName: parameter.spec.serviceName
            template: {
                metadata: labels: app: parameter.spec.template.metadata.labels.app
                spec: {
                    containers: [{
                        name: "nginx"
                        ports: [{
                            name:          "web"
                            containerPort: 80
                        }]
                        image: parameter.spec.template.spec.containers[0].image
                        volumeMounts: [{
                            name:      parameter.spec.template.spec.containers[0].volumeMounts[0].name
                            mountPath: parameter.spec.template.spec.containers[0].volumeMounts[0].mountPath
                        }]
                    }]
                    terminationGracePeriodSeconds: parameter.spec.template.spec.terminationGracePeriodSeconds 
                }
            }
            volumeClaimTemplates: [{
                metadata: name: parameter.spec.volumeClaimTemplates[0].metadata.name
                spec: {
                    accessModes: parameter.spec.volumeClaimTemplates[0].spec.accessModes 
                    resources: requests: storage: parameter.spec.volumeClaimTemplates[0].spec.resources.requests.storage 
                    storageClassName: parameter.spec.volumeClaimTemplates[0].spec.storageClassName 
                }
            }]
        }
    }
    outputs: {}
    parameter: {
        apiVersion: string | *"apps/v1"
        kind:       string | *"StatefulSet"
        metadata: name: string | *"web"
        spec: {
            replicas: int | *3
            serviceName: string | *"nginx"
            template: {
                metadata: labels: {app: string | *"nginx"}
                spec: {
                    containers: [{
                        name: string | *"nginx"
                        ports: [{
                            name:          string | *"web"
                            containerPort: int | *80
                        }]
                        image: string | *"k8s.gcr.io/nginx-slim:0.8"
                        volumeMounts: [{
                            name:      string | *"www"
                            mountPath: string | *"/usr/share/nginx/html"
                        }]
                    }]
                    terminationGracePeriodSeconds: int | *10
                }
            }
            volumeClaimTemplates: [{
                metadata: name: string | *"www"
                spec: {
                    accessModes: [string] | *["ReadWriteOnce"]
                    resources: requests: {storage: string | *"1Gi"}
                    storageClassName: string | *"my-storage-class"
                }
            }]
        }
    }
}
```
